Fix farmer rewards checks and add observer tests

emlowe opened the pull request:
The GUI farmer rewards check was not checking for observer keys and the code was duplicated from farmer and wallet, which is one reason this was missed.

Consolidated some code into new function called by both the wallet and farmer and added tests for observer keys.

Also added the ability to pass in the number of derivations to the RPC call - primarily this was so the tests can pass in much smaller numbers than the default (500) and so the tests can run much faster. However, this flexibility seems worth while regardless

wjblanke committed:
updated gui to d5b75bcf7a0fe1ef76775e7f1d5e12d169069676

emlowe committed:
Fix farmer rewards checks and add observer tests

emlowe committed:
Fix farmer rewards checks and add observer tests

xdustinface reviewed (changes requested):
I will review tomorrow but one thing i noticed as i skimmed it: I think you should drop all those unrelated `isort` changes.. imo we shouldn't just mix those into other PRs to have clear scopes why things have changed.

paninaro commented on chia/farmer/farmer.py:
```diff
@@ -500,7 +499,7 @@ async def update_pool_farmer_info() -> Tuple[Optional[GetFarmerResponse], Option
                         if error_code == PoolErrorCode.FARMER_NOT_KNOWN:
                             # Make the farmer known on the pool with a POST /farmer
-                            owner_sk_and_index: Optional[PrivateKey, uint32] = find_owner_sk(
+                            owner_sk_and_index: Optional[Tuple[G1Element, uint32]] = find_owner_sk(
```
Seems odd to me that `find_owner_sk` is returning a `G1Element` instead of a `PrivateKey`. I would fix the return type of `find_owner_sk` instead of changing the type of `owner_sk_and_index`.

emlowe replied:
I wasn't sure which direction to go here - I also found it odd, but that is long-standing code, so I wasn't sure.

emlowe replied:
Fixed, turns out all callers were expecting find_owner_sk to return PrivateKey - so fixed that

paninaro commented on chia/farmer/farmer.py:
```diff
@@ -524,7 +523,7 @@ async def update_pool_farmer_info() -> Tuple[Optional[GetFarmerResponse], Option
                         )
                         if payout_instructions_update_required or error_code == PoolErrorCode.INVALID_SIGNATURE:
-                            owner_sk_and_index: Optional[PrivateKey, uint32] = find_owner_sk(
+                            owner_sk_and_index: Optional[Tuple[G1Element, uint32]] = find_owner_sk(
```
Same comment here re: `find_owner_sk` return type.

paninaro commented on chia/farmer/farmer.py:
```diff
@@ -547,25 +546,30 @@ def get_public_keys(self):
+            have_farmer_sk, have_pool_sk = False, False
+            search_addresses: List[bytes32] = [self.farmer_target, self.pool_target]
+            for sk, _ in all_sks:
+                found_dict: Dict[bytes32, bool] = match_address_to_sk(sk, search_addresses, max_ph_to_search)
```
minor: A set would be a bit tidier to work with, but the dictionary approach works as well.

emlowe replied:
yea, good idea

paninaro commented on chia/wallet/derive_keys.py:
```diff
@@ -110,3 +112,26 @@ def find_authentication_sk(all_sks: List[PrivateKey], owner_pk: G1Element) -> Op
+
+def match_address_to_sk(
+    sk: PrivateKey, addresses_to_search: List[bytes32], max_ph_to_search: int = 500
+) -> Dict[bytes32, bool]:
```
minor: As mentioned earlier, a set would be a bit more natural to work with, but not critical.

paninaro commented on chia/wallet/derive_keys.py:
```diff
@@ -110,3 +112,26 @@ def find_authentication_sk(all_sks: List[PrivateKey], owner_pk: G1Element) -> Op
+            create_puzzlehash_for_pk(master_sk_to_wallet_sk(sk, uint32(i)).get_g1()),
+            create_puzzlehash_for_pk(master_sk_to_wallet_sk_unhardened(sk, uint32(i)).get_g1()),
+        ]
+        for address in addresses_to_search:
```
I'd make a copy of addresses_to_search and remove entries as they're found.

emlowe replied:
I thought about this, but at the moment, it's only used for a List of 2 and for small lists, I thought that overhead wouldn't be worth it. Although there is some problem here if the list contains any duplicated addresses, this will loop around more than required as the length check doesn't work in this case, so I'll need to fix that.

wjblanke committed:
Revert "derivation from just a master public key (#11140)" (#11143)

This reverts commit db536c615a0c7226ea8319fb98ec545eded23670.

cmmarslender committed:
Mark the github workspace as safe (#11159)

* Mark the github workspace as safe

* Move the git config step after git is installed in the test containers

emlowe committed:
unsort

emlowe committed:
unsort 2

emlowe committed:
Updated to use Sets

emlowe committed:
Callers of find_owner_sk were expecting PrivateKey

wjblanke committed:
updated gui to 81303fb962f4a627a2e1c55098e187a9057745da

arvidn committed:
optimize wallet tool by not caching the puzzle_hash -> derivation index, but caching puzzle_hash -> secret key (which is the lookup we're actually interested in). This avoids duplicating the actual derivation (#11154)

paninaro reviewed (dismissed):
lgtm

arvidn committed:
make listen port colissions in CI less likely (#11164)

cmmarslender committed:
Use get latest release endpoint for plotters, so that we ignore any pre-releases that could be returned by listReleases (#11165)

cmmarslender committed:
Build cli only version of debs (#11166)

* Build cli only version of debs

* Export the vars needed by j2

* Fix paths

* Add symlink to chia in /usr/local/bin/

* Upload the cli only debs to s3

* Add init.py

* Ensure SHA is on the dev build for amd64

arvidn committed:
add tool to generate a blockchain with full blocks, as a benchmark (#11146)

cmmarslender committed:
Simplify how the chia symlink is created in the CLI .deb (#11188)

arvidn committed:
fix block_tools feature when specifying a list of block references. Also add feature keep_going_until_tx_block. (#11185)

cmmarslender committed:
Fix filename of latest intel dev installer (#11203)

cmmarslender committed:
Add start_crawler and start_seeder to pyinstaller config (#11205)

cmmarslender committed:
Pin mac intel installer to 10.15 (#11209)

cmmarslender committed:
Revert "Pin mac intel installer to 10.15 (#11209)" (#11210)

This reverts commit 93a61eece1ab537f40cf3daf76be69dfce405762.

wallentx committed:
Adding changelog (#11223)

wallentx committed:
Adding changelog (#11223)

cmmarslender committed:
Revert "Revert "Pin mac intel installer to 10.15 (#11209)" (#11210)" (#11211)

This reverts commit d3e73a75ab44c799f8b6f9a76fab550ad6d7824a.

cmmarslender committed:
Fix install.sh test for bookworm (#11227)

altendky committed:
aiohttp==3.8.1 for Python 3.10 (#11129)

altendky committed:
replace some asyncio.get_event_loop().run_until_complete() with asyncio.run() (#11131)

altendky committed:
fix ssl context creation for server vs. client side (#11134)

xdustinface committed:
plotting|tests|setup: Improve `PlotManager` cache (#9903)

* plotting|tests: Extend the lifetime of plot data cache entries

* plotting|tests: Add `prover` and `farmer_public_key` to the cache

* plotting: Missing type hints in `Cache` and `CacheEntry`

* setup: Bump `chiapos` to 1.0.10

* plotting: Move cache classes into `chia.plotting.cache`

* plotting: Introduce `CacheEntry.from_disk_prover`

* Make `Cache` a dataclass

* `staticmethod` -> `classmethod`

* Store cache data as `bytes` in `DiskCache`

xdustinface committed:
streamable: Enable `isort` + more `mypy` (#10539)

* isort: Fix `streamable.py` and `test_streamable.py`

* mypy: Drop `streamable.py` and `test_streamable.py` form exclusion

And fix all the mypy issues.

* Fix `pylint`

* Introduce `ParseFunctionType` and `StreamFunctionType`

* Use `object` instead of `Type[Any]` for `is_type_*` functions

* Some `Any` -> `object`

* Use `typing.overload` for `recurse_jsonify`

* Move some comments

* Drop `Union`, use `Literal` properly

* Explicitly ignore the return of `f_type.parse`

Co-authored-by: Kyle Altendorf <sda@fstab.net>

* Merge two `recurse_jsonify` overloads

* Typing for the base definition of `recurse_jsonify`

Co-authored-by: Kyle Altendorf <sda@fstab.net>

fchirica committed:
Fix timelord closing. (#10630)

arvidn committed:
apply stricter mempool conditions (no garbage at the end of condition arguments). Fixup test (#10737)

xdustinface committed:
wallet: Fix `CATLineageStore` creation in `create_new_cat_wallet` (#10791)

* Implement `test_cat_creation_unique_lineage_store()`

Make sure creating CAT wallets leads to unique lineage stores.

* Fix `CATLineageStore` creation in `create_new_cat_wallet`

We currently create the `CATLineageStore` before the new asset id exists
which leads to all CATs created by `create_new_cat_wallet` using the
same lineage store table:
`lineage_proofs_0000000000000000000000000000000000000000`.

* Create `CATLineageStore` in `generate_issuance_bundle`

altendky committed:
--module pytest (#10940)

Commit:
Fix incorrect return in "rollback_to_block" (#10954)

* Add comments and fix a coin subscription bug

* Updated rollback test

* simplify test

* expand test

Co-authored-by: Mariano Sorgente <sorgente711@gmail.com>

altendky committed:
Correct wallet CLI sent/received indication (#11006)

* Correct wallet CLI sent/received indication

* extract dict to reusable global scope, plus a function

xdustinface committed:
farmer: Introduce `UPDATE_POOL_INFO_FAILURE_RETRY_INTERVAL` (#11076)

Retry to load the pool info with a 2 minutes interval if it failed.

emlowe committed:
Expose transaction sorting options to CLI (#11090)

mariano54 committed:
Optimize test_wallet_sync (#11103)

* Optimize test_wallet_sync

* Update tests/wallet/sync/test_wallet_sync.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

* Fix test

Co-authored-by: Kyle Altendorf <sda@fstab.net>

altendky committed:
Up Windows installer build timeout to 50 minutes (#11107)

Recent runs are mostly low 30s but some are close to 40 and one just timed out.

xdustinface committed:
wallet: Drop unused `WalletStateManager.get_derivation_index` (#11110)

altendky committed:
locate pyinstaller in dev deps rather than build scripts (#11118)

altendky committed:
remove setting of SCM_VERSION in install-version.py (#11119)

altendky committed:
Remove code related to no-longer-used websockets library (#11123)

* Remove logging tweak for no-longer-used websockets library

Follow up after https://github.com/Chia-Network/chia-blockchain/pull/10611.

* Update pylintrc

* Update test_daemon.py

xdustinface committed:
harvester: Enable `isort` (#11127)

xdustinface committed:
farmer: Enable `isort` (#11128)

aqk committed:
Correct "Older block not found" error message (#11130)

aqk committed:
Print MempoolInclusionStatus as string when reporting mempool inclusion status (#11133)

dependabot[bot] committed:
Bump actions/upload-artifact from 2 to 3 (#11144)

* Bump actions/upload-artifact from 2 to 3

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

* Also update runner_templates

* Mark the github workspace as safe (#11159)

* Mark the github workspace as safe

* Move the git config step after git is installed in the test containers

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Gene Hoffman <hoffmang@hoffmang.com>
Co-authored-by: Chris Marslender <chrismarslender@gmail.com>

xdustinface committed:
wallet: Some rollback fixes (#11149)

* Set `in_transaction` for `set_finished_sync_up_to`

* Call `clean_block_records` based on `in_transaction`

* Don't pop wallets in `reorg_rollback`, do it after it was commited

arvidn committed:
optimize Program.curry() (#11162)

* optimize Program.curry()

* Add a comment for this curry implementation

* optimize Program.curry()

* Add comment

Co-authored-by: arty <art.yerkes@gmail.com>

xdustinface committed:
harvester: Use a set instead of a list to speed up availability checks (#11204)

xdustinface committed:
cmds: Only call `is_running` endpoint once in `async_start` (#11221)

hoffmang9 committed:
Remove 3.9.11 for Windows Installer (#11226)

And replace with just 3.9.x

trepca committed:
Use log formatter to avoid serialisation (#11184)

* avoid serialisation in logging

* avoid serialisation in logging

* Revert "avoid serialisation in logging"

This reverts commit 4a2b4473964ecf0863ae6860e72d54a4f5d8958f.

aqk committed:
Improve detection of disconnected websocket between services (#11069)

wjblanke committed:
updated bls and vdf to new releases

xdustinface committed:
github: Fix `plot_sync` workflows after #10940 merge (#11241)

cmmarslender committed:
Add cli only rpm (#11236)

* Add cli only rpm

* Ensure rvm (fpm) is loaded before running fpm

* Use full path to fpm, since GHA seems to mess up the PATH in the container

* Add back source and add use ruby-3

* Call rpm script with bash, to see if the rvm script works

* Add --depends for libcrypt.so.1
